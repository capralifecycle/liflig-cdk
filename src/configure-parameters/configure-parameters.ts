import * as iam from "@aws-cdk/aws-iam"
import * as secretsmanager from "@aws-cdk/aws-secretsmanager"
import * as ssm from "@aws-cdk/aws-ssm"
import * as cdk from "@aws-cdk/core"
import * as crypto from "crypto"

/**
 * A plain text application parameter that will be stored
 * in AWS Parameter Store. Only used for non-sensitive values,
 * typically those commited directly in the IaC code.
 */
export interface PlainTextParameter {
  key: string
  value: string
}

/**
 * An AWS Secret that should hold a JSON object.
 *
 * This will provide a secret that liflig-properties can read. The final
 * parameter seen by the application will be the key given here and the
 * keys from the secret JSON object appended.
 */
export interface SecretParameter {
  key: string
  secret: secretsmanager.ISecret
}

/**
 * An AWS Secret that should hold a JSON object.
 *
 * This will provide a secret that liflig-properties can read. The final
 * parameter seen by the application will be the key given here and the
 * keys from the secret JSON object appended.
 */
export interface SecretByNameParameter {
  key: string
  secretName: string
}

export type Parameter =
  | PlainTextParameter
  | SecretParameter
  | SecretByNameParameter

interface Props {
  /**
   * Prefix used for parameter names.
   * Should start with '/' and end without '/'.
   */
  ssmPrefix: string
  parameters: Parameter[]
}

export class ConfigureParameters {
  public parameters: ssm.IParameter[]
  public ssmPrefix: string
  private secrets: secretsmanager.ISecret[]

  private configParameters: PlainTextParameter[]
  private secretParameters: SecretParameter[]

  constructor(scope: cdk.Construct, props: Props) {
    if (!props.ssmPrefix.startsWith("/") || props.ssmPrefix.endsWith("/")) {
      throw new Error("ssmPrefix should start with '/' and end without '/'")
    }

    this.ssmPrefix = props.ssmPrefix

    this.configParameters = []
    this.secretParameters = []

    for (const parameter of props.parameters) {
      if ("value" in parameter) {
        this.configParameters.push(parameter)
      } else if ("secret" in parameter) {
        this.secretParameters.push(parameter)
      } else if ("secretName" in parameter) {
        this.secretParameters.push({
          key: parameter.key,
          secret: secretsmanager.Secret.fromSecretNameV2(
            scope,
            `SecretRef${parameter.key}`,
            parameter.secretName,
          ),
        })
      }
    }

    const result: ssm.IParameter[] = []

    this.configParameters.forEach((it) => {
      const param = new ssm.StringParameter(scope, `Config${it.key}`, {
        type: ssm.ParameterType.STRING,
        stringValue: it.value,
        parameterName: `${this.ssmPrefix}/config/${it.key}`,
      })
      result.push(param)
    })

    this.secretParameters.forEach((it) => {
      const param = new ssm.StringParameter(scope, `Secret${it.key}`, {
        type: ssm.ParameterType.STRING,
        stringValue: it.secret.secretArn,
        parameterName: `${this.ssmPrefix}/secrets/${it.key}`,
      })
      result.push(param)
    })

    this.secrets = this.secretParameters.map((it) => it.secret)

    this.parameters = result
  }

  grantRead(grantable: iam.IGrantable): void {
    grantable.grantPrincipal.addToPrincipalPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ["ssm:GetParametersByPath"],
        resources: [
          `arn:aws:ssm:${cdk.Stack.of(grantable).region}:${
            cdk.Stack.of(grantable).account
          }:parameter${this.ssmPrefix}/*`,
        ],
      }),
    )

    for (const secret of this.secrets) {
      secret.grantRead(grantable)
    }
  }

  /**
   * Produce a checksum-value that can be used as a kind of
   * nonce to trigger redeployment on parameter changes.
   *
   * Note: If the parameter references a token no change will be visible,
   * and a manual redeployment might be needed.
   */
  get hashValue(): string {
    const hash = crypto.createHash("sha256")

    if (this.configParameters) {
      this.configParameters.forEach((it) => {
        hash.update("|")
        hash.update(it.key)
        hash.update("|")

        if (!cdk.Token.isUnresolved(it.value)) {
          hash.update(it.value)
        }
      })
    }

    if (this.secretParameters) {
      this.secretParameters.forEach((it) => {
        hash.update("|")
        hash.update(it.key)
        hash.update("|")

        if (!cdk.Token.isUnresolved(it.secret.secretArn)) {
          hash.update(it.secret.secretArn)
        }
      })
    }

    return hash.digest("hex")
  }
}
