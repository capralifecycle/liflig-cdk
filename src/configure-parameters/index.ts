export {
  ConfigureParameters,
  ConfigureParametersProps,
  PlainTextParameter,
  SecretParameter,
  SecretByNameParameter,
  Parameter,
} from "./configure-parameters"
