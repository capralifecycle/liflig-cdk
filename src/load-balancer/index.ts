export { LoadBalancer } from "./load-balancer"
export type { LoadBalancerProps } from "./load-balancer"
