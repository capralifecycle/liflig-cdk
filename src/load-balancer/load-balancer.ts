import * as certificatemanager from "@aws-cdk/aws-certificatemanager"
import * as ec2 from "@aws-cdk/aws-ec2"
import * as elb from "@aws-cdk/aws-elasticloadbalancingv2"
import { ListenerAction } from "@aws-cdk/aws-elasticloadbalancingv2"
import * as s3 from "@aws-cdk/aws-s3"
import * as cdk from "@aws-cdk/core"

export interface LoadBalancerProps {
  certificates: certificatemanager.ICertificate[]
  vpc: ec2.IVpc
  overrideLoadBalancerProps?: Partial<elb.ApplicationLoadBalancerProps>
}

export class LoadBalancer extends cdk.Construct {
  public readonly loadBalancer: elb.ApplicationLoadBalancer
  public readonly httpsListener: elb.ApplicationListener
  public readonly accessLogsBucket: s3.Bucket

  constructor(scope: cdk.Construct, id: string, props: LoadBalancerProps) {
    super(scope, id)

    this.loadBalancer = new elb.ApplicationLoadBalancer(this, "LoadBalancer", {
      vpc: props.vpc,
      internetFacing: true,
      vpcSubnets: props.vpc.selectSubnets({
        subnetType: ec2.SubnetType.PUBLIC,
      }),
      ...props.overrideLoadBalancerProps,
    })

    this.loadBalancer
      .addListener("HttpListener", {
        port: 80,
      })
      .addAction("HttpsRedirect", {
        action: ListenerAction.redirect({
          port: "443",
          protocol: "HTTPS",
          permanent: true,
        }),
      })

    // The Load Balancer require a default target group.
    // We will not connect anything to the default target group.
    const defaultTargetGroup = new elb.ApplicationTargetGroup(
      this,
      "DefaultTargetGroup",
      {
        protocol: elb.ApplicationProtocol.HTTP,
        port: 80,
        vpc: props.vpc,
        targetType: elb.TargetType.INSTANCE,
      },
    )

    this.httpsListener = this.loadBalancer.addListener("HttpsListener", {
      sslPolicy: elb.SslPolicy.TLS12,
      protocol: elb.ApplicationProtocol.HTTPS,
      port: 443,
      certificates: props.certificates,
      defaultTargetGroups: [defaultTargetGroup],
    })

    this.accessLogsBucket = new s3.Bucket(this, "AccessLogsBucket", {
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
      lifecycleRules: [
        {
          expiration: cdk.Duration.days(30),
        },
      ],
    })

    this.loadBalancer.logAccessLogs(this.accessLogsBucket)
  }
}
