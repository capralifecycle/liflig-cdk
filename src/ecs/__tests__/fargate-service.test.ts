import "@aws-cdk/assert/jest"
import * as cm from "aws-cdk-lib/aws-certificatemanager"
import * as ec2 from "aws-cdk-lib/aws-ec2"
import * as ecr from "aws-cdk-lib/aws-ecr"
import * as ecs from "aws-cdk-lib/aws-ecs"
import * as route53 from "aws-cdk-lib/aws-route53"
import * as sm from "aws-cdk-lib/aws-secretsmanager"
import { App, Stack } from "aws-cdk-lib"
import "jest-cdk-snapshot"
import { FargateService, ListenerRule } from ".."
import { Parameter } from "../../configure-parameters/configure-parameters"
import { LoadBalancer } from "../../load-balancer"

test("creates fargate service with parameters and listener rule", () => {
  const app = new App()
  const supportStack = new Stack(app, "SupportStack", {
    env: {
      region: "eu-west-1",
    },
  })
  const stack = new Stack(app, "Stack", {
    env: {
      region: "eu-west-1",
    },
  })

  const vpc = new ec2.Vpc(supportStack, "Vpc")

  const hostedZone = new route53.HostedZone(supportStack, "HostedZone", {
    zoneName: "example.com",
  })

  const certificate = new cm.Certificate(supportStack, "Certificate", {
    domainName: `*.example.com`,
    subjectAlternativeNames: ["example.com"],
    validation: cm.CertificateValidation.fromDns(hostedZone),
  })

  const loadBalancer = new LoadBalancer(supportStack, "LoadBalancer", {
    certificates: [certificate],
    vpc: vpc,
  })

  const ecsCluster = new ecs.Cluster(supportStack, "Cluster", {
    vpc,
  })

  const ecrRepository = new ecr.Repository(supportStack, "Repository", {
    repositoryName: "example-repository",
  })

  const parameters: Parameter[] = [
    {
      key: "ExamplePlainTextParameter",
      value: "ExamplePlainTextParameter",
    },
    {
      key: "ExampleSecretParameter",
      secret: sm.Secret.fromSecretNameV2(
        stack,
        "ExampleSecretParameter",
        "ExampleSecretParameter",
      ),
    },
    {
      key: "ExampleSecretByNameParameter",
      secretName: "ExampleSecretByNameParameter",
    },
  ]

  const service = new FargateService(stack, "Service", {
    serviceName: "example-service",
    vpc: vpc,
    cluster: ecsCluster,
    desiredCount: 2,
    parameters,
    ecsImage: ecs.ContainerImage.fromEcrRepository(
      ecrRepository,
      "exampleEcrTag",
    ),
    enableCircuitBreaker: true,
  })

  new ListenerRule(stack, "Dns", {
    domainName: `example.com`,
    hostedZone: hostedZone,
    httpsListener: loadBalancer.httpsListener,
    listenerPriority: 10,
    loadBalancer: loadBalancer.loadBalancer,
    targetGroup: service.targetGroup!,
  })

  expect(stack).toMatchCdkSnapshot()
})
