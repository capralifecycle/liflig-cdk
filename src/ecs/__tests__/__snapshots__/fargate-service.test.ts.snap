// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates fargate service with parameters and listener rule 1`] = `
Object {
  "Resources": Object {
    "DnsARecordFED6ED67": Object {
      "Properties": Object {
        "AliasTarget": Object {
          "DNSName": Object {
            "Fn::Join": Array [
              "",
              Array [
                "dualstack.",
                Object {
                  "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttLoadBalancer8E47CE52DNSName59BE487A",
                },
              ],
            ],
          },
          "HostedZoneId": Object {
            "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttLoadBalancer8E47CE52CanonicalHostedZoneID7C0E7DD7",
          },
        },
        "HostedZoneId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefHostedZoneDB99F8662BBAE844",
        },
        "Name": "example.com.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "DnsListenerRule18B182CC": Object {
      "Properties": Object {
        "Actions": Array [
          Object {
            "TargetGroupArn": Object {
              "Ref": "ServiceTargetGroupE74A3394",
            },
            "Type": "forward",
          },
        ],
        "Conditions": Array [
          Object {
            "Field": "host-header",
            "Values": Array [
              "example.com",
            ],
          },
        ],
        "ListenerArn": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefLoadBalancerHttpsListener68D81C9458BEEF85",
        },
        "Priority": 10,
      },
      "Type": "AWS::ElasticLoadBalancingV2::ListenerRule",
    },
    "Service9571FDD8": Object {
      "DependsOn": Array [
        "DnsListenerRule18B182CC",
        "ServiceConfigExamplePlainTextParameterC948638B",
        "ServiceSecretExampleSecretByNameParameterB153ADD4",
        "ServiceSecretExampleSecretParameter2147C87C",
      ],
      "Properties": Object {
        "Cluster": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefClusterEB0386A796A0E3FE",
        },
        "DeploymentConfiguration": Object {
          "MaximumPercent": 200,
          "MinimumHealthyPercent": 100,
        },
        "DesiredCount": 2,
        "EnableECSManagedTags": false,
        "EnableExecuteCommand": true,
        "HealthCheckGracePeriodSeconds": 60,
        "LaunchType": "FARGATE",
        "LoadBalancers": Array [
          Object {
            "ContainerName": "Container",
            "ContainerPort": 8080,
            "TargetGroupArn": Object {
              "Ref": "ServiceTargetGroupE74A3394",
            },
          },
        ],
        "NetworkConfiguration": Object {
          "AwsvpcConfiguration": Object {
            "AssignPublicIp": "ENABLED",
            "SecurityGroups": Array [
              Object {
                "Fn::GetAtt": Array [
                  "ServiceSecurityGroupC96ED6A7",
                  "GroupId",
                ],
              },
            ],
            "Subnets": Array [
              Object {
                "Fn::ImportValue": "SupportStack:ExportsOutputRefVpcPublicSubnet1Subnet5C2D37C4FFA2B456",
              },
              Object {
                "Fn::ImportValue": "SupportStack:ExportsOutputRefVpcPublicSubnet2Subnet691E08A351552740",
              },
            ],
          },
        },
        "PlatformVersion": "1.4.0",
        "ServiceName": "example-service",
        "TaskDefinition": Object {
          "Ref": "ServiceTaskDefinition55FA0F15",
        },
      },
      "Type": "AWS::ECS::Service",
    },
    "ServiceConfigExamplePlainTextParameterC948638B": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/config/ExamplePlainTextParameter",
        "Type": "String",
        "Value": "ExamplePlainTextParameter",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceLogGroupB910EE76": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "RetentionInDays": 14,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "ServiceSecretExampleSecretByNameParameterB153ADD4": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/secrets/ExampleSecretByNameParameter",
        "Type": "String",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:eu-west-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":secret:ExampleSecretByNameParameter",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceSecretExampleSecretParameter2147C87C": Object {
      "Properties": Object {
        "Name": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/secrets/ExampleSecretParameter",
        "Type": "String",
        "Value": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:eu-west-1:",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":secret:ExampleSecretParameter",
            ],
          ],
        },
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ServiceSecurityGroupC96ED6A7": Object {
      "Properties": Object {
        "GroupDescription": "Stack/Service/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefVpc8378EB38272D6E3A",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "ServiceSecurityGroupSupportStackLoadBalancerSecurityGroup43D42DD38080from8AFD9D71": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "FromPort": 8080,
        "GroupId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttLoadBalancerSecurityGroupC89CA14BGroupId6F8F4447",
        },
        "IpProtocol": "tcp",
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "ServiceSecurityGroupfromSupportStackLoadBalancerSecurityGroup43D42DD3808010A4CCEF": Object {
      "Properties": Object {
        "Description": "Load balancer to target",
        "FromPort": 8080,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "ServiceSecurityGroupC96ED6A7",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttLoadBalancerSecurityGroupC89CA14BGroupId6F8F4447",
        },
        "ToPort": 8080,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "ServiceTargetGroupE74A3394": Object {
      "Properties": Object {
        "HealthCheckIntervalSeconds": 10,
        "HealthCheckPath": "/health",
        "HealthyThresholdCount": 2,
        "Port": 8080,
        "Protocol": "HTTP",
        "TargetGroupAttributes": Array [
          Object {
            "Key": "deregistration_delay.timeout_seconds",
            "Value": "15",
          },
          Object {
            "Key": "stickiness.enabled",
            "Value": "false",
          },
        ],
        "TargetType": "ip",
        "VpcId": Object {
          "Fn::ImportValue": "SupportStack:ExportsOutputRefVpc8378EB38272D6E3A",
        },
      },
      "Type": "AWS::ElasticLoadBalancingV2::TargetGroup",
    },
    "ServiceTaskDefinition55FA0F15": Object {
      "Properties": Object {
        "ContainerDefinitions": Array [
          Object {
            "Environment": Array [
              Object {
                "Name": "SSM_PREFIX",
                "Value": "/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters",
              },
              Object {
                "Name": "PARAMS_HASH",
                "Value": "90fa2bdb2b8eb3d9ec2b313224ba0c78d3905bd65f6d1bc6427158370196cdb8",
              },
            ],
            "Essential": true,
            "Image": Object {
              "Fn::Join": Array [
                "",
                Array [
                  Object {
                    "Fn::Select": Array [
                      4,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  Object {
                    "Fn::Select": Array [
                      3,
                      Object {
                        "Fn::Split": Array [
                          ":",
                          Object {
                            "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  Object {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  Object {
                    "Fn::ImportValue": "SupportStack:ExportsOutputRefRepository22E53BBD9A9E013B",
                  },
                  ":exampleEcrTag",
                ],
              ],
            },
            "LinuxParameters": Object {
              "Capabilities": Object {},
              "InitProcessEnabled": true,
            },
            "LogConfiguration": Object {
              "LogDriver": "awslogs",
              "Options": Object {
                "awslogs-datetime-format": "%Y-%m-%dT%H:%M:%S",
                "awslogs-group": Object {
                  "Ref": "ServiceLogGroupB910EE76",
                },
                "awslogs-region": "eu-west-1",
                "awslogs-stream-prefix": "ecs",
              },
            },
            "Name": "Container",
            "PortMappings": Array [
              Object {
                "ContainerPort": 8080,
                "HostPort": 8080,
                "Protocol": "tcp",
              },
            ],
          },
        ],
        "Cpu": "256",
        "ExecutionRoleArn": Object {
          "Fn::GetAtt": Array [
            "ServiceTaskDefinitionExecutionRole5D3DB197",
            "Arn",
          ],
        },
        "Family": "StackServiceTaskDefinition81B5CCD5",
        "Memory": "512",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": Array [
          "FARGATE",
        ],
        "TaskRoleArn": Object {
          "Fn::GetAtt": Array [
            "ServiceTaskDefinitionTaskRole3BD32B0F",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "ServiceTaskDefinitionExecutionRole5D3DB197": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceTaskDefinitionExecutionRoleDefaultPolicyE16C69DE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "SupportStack:ExportsOutputFnGetAttRepository22E53BBDArn3AD4E30B",
              },
            },
            Object {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "ServiceLogGroupB910EE76",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceTaskDefinitionExecutionRoleDefaultPolicyE16C69DE",
        "Roles": Array [
          Object {
            "Ref": "ServiceTaskDefinitionExecutionRole5D3DB197",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ServiceTaskDefinitionTaskRole3BD32B0F": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ServiceTaskDefinitionTaskRoleDefaultPolicyED991BA3": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/liflig-cdk/Stack/c8cabff989df74f9b2755f4ae923cc6492f944970b/parameters/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:ExampleSecretParameter-??????",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":secretsmanager:eu-west-1:",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":secret:ExampleSecretByNameParameter-??????",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssmmessages:CreateControlChannel",
                "ssmmessages:CreateDataChannel",
                "ssmmessages:OpenControlChannel",
                "ssmmessages:OpenDataChannel",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "logs:DescribeLogGroups",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogStream",
                "logs:DescribeLogStreams",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ServiceTaskDefinitionTaskRoleDefaultPolicyED991BA3",
        "Roles": Array [
          Object {
            "Ref": "ServiceTaskDefinitionTaskRole3BD32B0F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
